### PR TITLE
Additional type hints for client REST servlets (part 3)

### PR DESCRIPTION
**Pull request by clokep**
Follow-up to #10665/#10674 with more modules.

~~Built-on #10706.~~

### CODE REVIEW

File: synapse/rest/client/register.py [hunk 1]
```diff
@@ -445,23 +429,21 @@ def __init__(self, hs):
...
 
-        if kind == b"guest":
+        if kind == "guest":
             ret = await self._do_guest_registration(body, address=client_addr)
             return ret
-        elif kind != b"user":
+        elif kind != "user":
             raise UnrecognizedRequestError(
-                "Do not understand membership kind: %s" % (kind.decode("utf8"),)
+                "Do not understand membership kind: %s" % (kind,)
```
Comment (erikjohnston) [inline synapse/rest/client/register.py:446]:
```suggestion
                f"Do not understand membership kind: {kind}",
```

while we're here

File: synapse/rest/client/room.py [hunk 2]
```diff
@@ -929,48 +1004,58 @@ async def on_POST(self, request):
...
-    if with_get:
+    on_POST = getattr(servlet, "on_POST", None)
+    if on_POST:
+        http_server.register_paths(
+            "POST",
+            client_patterns(regex_string + "$", v1=True),
+            on_POST,
+            servlet.__class__.__name__,
+        )
+    on_PUT = getattr(servlet, "on_PUT", None)
```
Comment (erikjohnston) [inline synapse/rest/client/room.py:1045]: Do we want to allow servlet classes to not have a `PUT` here? This feels more like we should be asserting here instead (which will cause Synapse not to start)

**Comment by erikjohnston** [comment]
LGTM otherwise

File: synapse/rest/client/room.py (hunk 2 above)
Comment (clokep) [reply synapse/rest/client/room.py:1045]: That's fine. I figured we can go either way. Annoyingly an mypy doesn't seem to work great with `hasattr`. 😢 I'll take another look!

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**